csvcubed/csvcubed/constraints/columns.py

```diff
@@ -0,0 +1,219 @@
+from dataclasses import dataclass
+from enum import Enum
+import json
+from typing import Dict, List, Optional
+
+
+class ColType(Enum):
+    DIMENSION = "dimension"
+    MEASURE = "measure"
+    ATTRIBUTE = "attribute"
+    MEASURE_TYPE = "measure type"
+
+
+@dataclass
+class ColumnRef:
+    """
+    A class to hold the information from the [tables][tableSchema] dictionary of the csvw
+    defining a column, along with additional information that can inferred from that.
+    """
+
+    csvw_as_dict: Optional[dict]  # pointer
+    initial_column_dict: dict
+
+    # (possible) existing values
+    name: Optional[str]
+    title: Optional[str]
+    required: Optional[str] = None
+    propertyUrl: Optional[str] = None
+    valueUrl: Optional[str] = None
+
+    # values we (possibly) are inferring
+    implicit_component_postfix: Optional[str] = None
+    implicit_url_differentiator: Optional[str] = None
+    implied_type: Optional[ColType] = None
+
+    def serialised(self) -> str:
+        """
+        Simple json serialiser for debugging
+        """
+        jsonified = {
+            "from_csvw_column_definition": {
+                "name": self.name,
+                "title": self.title,
+                "required": self.required,
+                "propertyUrl": self.propertyUrl,
+                "valueUrl": self.valueUrl,
+            },
+            "implied_extrapolated_values": {
+                "implicit_component_postfix": self.implicit_component_postfix,
+                "implicit_url_differentiator": self.implicit_url_differentiator,
+                "implied_type": self.implied_type.value,
+            },
+        }
+
+        return json.dumps(jsonified, indent=2)
+
+    def populate(self):
+        """
+        Populates fields we can infer but don't have to begin with
+        """
+        self.set_implied_type()
+        self.set_implicit_component_postfix()
+        self.set_impicit_url_differentiator()
+
+    # TODO - consider case of not having a propertyUrl
+    def set_implied_type(self):
+        """
+        From the information we have, "guess" the type of column based on the
+        information provided within a given columns propertyUrl
+        """
+        if not self.propertyUrl:
+            return  # TODO - consider ramifications
+
+        matches = []
+
+        if any(["/dimension/" in self.propertyUrl, "/dimension#" in self.propertyUrl]):
```

Comment:
> cheers for looking at this, I know the code is a messy scramble of ideas. Good thoughts and a great point on leaning on the sparql queries more, ty 👍
> 
> I've got a few questions on things I need to clarify, just want to capture them while I'm reading
> 
>     * do we want this to be internal to csvcubed or its own thing? Installing pandas and a database is quite heavy duty to read rows, but obviously much less of a consideration if we're doing it anyway in csvcubed.
> 

So ideally it would be part of csvcubed, I imagine a command called `test` which could take a CSV-W input and run all of the tests we have against it (along with the validations we do when the user builds a cube).

I think pandas has been fine for what we've been working with so far but I think the complexity of the SPARQL queries combined with potentially large file sizes means using sqlite for this would probably be a better route to go down. I think the sqlite dependency is only a few mb tbh and is [installed by default with python](https://www.tutorialspoint.com/sqlite/sqlite_python.htm). 

> 
> I guess I'm thinking about one of the questions about whether we need to use pandas in csvcubed, no horse in that race, but it plays into this a bit. I was if anything thinking I'd be taking the pandas reliance out. I didn't do anything here that actually needs a dataframe, at least in so far as i went (though I didn't get to deduplication, that feels like the obvious challenging bit).
> 

I think finding duplicates should be quite simple with an appropriate sqlite table and query. I have certainly written Microsoft-SQL queries that highlight duplicate rows in the past so I'm fairly confident a similar query can be written in sqlite; you just have to group by the dimension + measure columns and return rows `HAVING COUNT(obs_val)  > 1`.

>     * are we only concerned with csvcubed produced csvw's? 

Yes. This is one of the muddy things with our project at the moment that people are talking about a scale of CSV-W maturity. Honestly, I think our primary purpose now is advocating for csvcubed or nothing. There's basically no way we can cope with all of the different ways people could possibly build CSV-Ws to represent the same information, so I'm planning on standardising on a format akin to what we're producing now so we can more easily write tools which extract the information reliably.

> It's still a bit muddy for me, so say csvcubed can never do X, it feels odd to validate that nobody has done X. 

This is a bit of a muddy one, yes. The problem is that users can still manually edit the CSVs and CSV-Ws and cause problem X to happen, so I think it's a good idea (if lower priority) to be able to highlight those problems.

> It feels related to initial question as if its general csvw checking - its its own thing - dependency considerations come into play, if that makes sense,